src/sage/rings/padics/witt_vector_ring.py

```diff
@@ -0,0 +1,917 @@
+
+        EXAMPLES::
+
+            sage: P.<X1,X2,Y1,Y2> = PolynomialRing(GF(3),'X1,X2,Y1,Y2')
+            sage: W = WittVectorRing(P, p=3, prec=2, algorithm='standard')
+            sage: W([X1,X2]) + W([Y1,Y2])  # indirect doctest
+            (X1 + Y1, -X1^2*Y1 - X1*Y1^2 + X2 + Y2)
+            sage: W([X1,X2]) * W([Y1,Y2])  # indirect doctest
+            (X1*Y1, X2*Y1^3 + X1^3*Y2)
+        """
+        p = self._prime
+        prec = self._prec
+        x_var_names = [f'X{i}' for i in range(prec)]
+        y_var_names = [f'Y{i}' for i in range(prec)]
+        var_names = x_var_names + y_var_names
+
+        # Okay, what's going on here? Sage, by default, relies on
+        # Singular for Multivariate Polynomial Rings, but Singular uses
+        # only SIXTEEN bits (unsigned) to store its exponents. So if we
+        # want exponents larger than 2^16 - 1, we have to use the
+        # generic implementation. However, after some experimentation,
+        # it seems like the generic implementation is faster?
+        #
+        # After trying to compute S_4 for p=5, it looks like generic is
+        # faster for  very small polys, and MUCH slower for large polys.
+        # So we'll default to singular unless we can't use it.
+        #
+        # Remark: Since when is SIXTEEN bits sufficient for anyone???
+        #
+        if p**(prec-1) >= 2**16:
+            implementation = 'generic'
+        else:
+            implementation = 'singular'
+
+        # We first generate the "universal" polynomials and then project
+        # to the coefficient ring.
+        R = PolynomialRing(ZZ, var_names, implementation=implementation)
+        x_y_vars = R.gens()
+        x_vars = x_y_vars[:prec]
+        y_vars = x_y_vars[prec:]
+
+        self._sum_polynomials = [0]*(self._prec)
+        for n in range(prec):
+            s_n = x_vars[n] + y_vars[n]
+            for i in range(n):
+                s_n += ((x_vars[i]**(p**(n-i)) + y_vars[i]**(p**(n-i))
+                        - self._sum_polynomials[i]**(p**(n-i))) / p**(n-i))
+            self._sum_polynomials[n] = R(s_n)
+
+        self._prod_polynomials = [x_vars[0] * y_vars[0]] + [0]*(self._prec-1)
+        for n in range(1, prec):
+            x_poly = sum([p**i * x_vars[i]**(p**(n-i)) for i in range(n+1)])
+            y_poly = sum([p**i * y_vars[i]**(p**(n-i)) for i in range(n+1)])
+            p_poly = sum([p**i * self._prod_polynomials[i]**(p**(n-i))
+                         for i in range(n)])
+            p_n = (x_poly*y_poly - p_poly) // p**n
+            self._prod_polynomials[n] = p_n
+
+        # We have to use generic here, because Singular doesn't support
+        # Polynomial Rings over Polynomial Rings. For example,
+        # ``PolynomialRing(GF(5)['x'], ['X', 'Y'],
+        #                  implementation='singular')``
+        # will fail.
+        S = PolynomialRing(self._coefficient_ring, x_y_vars,
+                           implementation='generic')
```

Comment:
```suggestion
        S = PolynomialRing(self._coefficient_ring, x_y_vars)
```
I'm not sure we really want to specify the implementation here; I think it's preferable to let Sage do as it prefers.

Reply:
This is legacy code, I guess there was a problem with previous versions, but indeed removing it runs fine. I made another commit because the comment was also outdated.